### DISCUSSION TITLE
Why does my query require a COLLECTION_GROUP_ASC index?

### DISCUSSION THREAD

**Discussion by pndalal** [GodotNuts/GodotFirebase#234] [Q&A] [answered]
I'm trying to perform a simple query that looks like this:
```
query.from("questions")
query.where("text", FirestoreQuery.OPERATOR.EQUAL, "What's your name?")
```

I get a Firebase error saying I need a collection group index. Why? Isn't this a basic query that is supported without adding any additional indexes? The _questions_ collection is top level, and the _text_ field is a top-level field for my documents.

**EDIT:**  I think I figured it out. There is an optional parameter for the "from" method called "all_descendants". It is true by default. I set it to false, and I no longer require the index, as expected. Here is the working code:

```
query.from("questions", false)
query.where("text", FirestoreQuery.OPERATOR.EQUAL, "What's your name?")
```

**Comment by WolfgangSenff** [comment]
Not anymore, unfortunately. This is a Firebase requirement now: all queries require indexes.

**Reply by pndalal** [reply]
Thanks for your reply. It's probably not a big deal for me either way, but according to [this](https://firebase.google.com/docs/firestore/query-data/index-overview?authuser=0#automatic_indexing), Firestore "automatically maintains single-field indexes for each field in a document", so according to that I believe I shouldn't have to create a manual index using a link. Also, on the single field index settings page I see an "Automatic Index Settings", which seems to cover the query in my example.

**Reply by WolfgangSenff** [reply]
The last point is:
`Single-field indexes with collection group scope are not maintained by default.`

My assumption when you wrote collection_group_asc was that it fell into that category. Is that not the case? Did your automatic index settings fix it? Very curious! I think I'm thinking about real-time database indexes, on second thought, regarding all searches/orders require indexes.

**Reply by pndalal** [reply]
No I don't believe that to be the case, because "questions" in my example is a single collection, of which I wish to query a single field. But I'm new to Firestore so perhaps I've misconfigured or misunderstood something.

What I meant to say about the Automatic Index Settings is that by default they seem to support my single collection scope query. I shouldn't have to make any changes, since I'm not attempting to query a collection group. Below is my entire query code:

```
var query : FirestoreQuery = FirestoreQuery.new()
query.from("questions")
query.where("text", FirestoreQuery.OPERATOR.EQUAL, "What's your name?")
var task : FirestoreTask = Firebase.Firestore.query(query)
var result = yield(task, "task_finished")
```

I have not specified anything to suggest I want to perform a collection group query. The only thing notable about my data, perhaps, is that within the questions collection, each document contains a sub-collection called "responses." I wonder if this has something to do with it.

EDIT: I tried to query a single field of another collection, "users", that has no nested sub-collections, and it still says it requires a COLLECTION_GROUP_ASC index. I don't know why.

**Reply by pndalal** [reply]
I think I figured it out. All-Descendants had to be set to false. By default it is true.

```
var query : FirestoreQuery = FirestoreQuery.new()
query.from("questions", false) // <-------- THIS LINE
query.where("text", FirestoreQuery.OPERATOR.EQUAL, "What's your name?")
var task : FirestoreTask = Firebase.Firestore.query(query)
var result = yield(task, "task_finished")
```

**Comment by fenix-hub** [comment, marked as answer]
@pndalal Hi,
I'm noticing that you are using two times the `.from` function.
I suppose that your `.from("questions")` is correct, but to use an operator you have to issue the `.where()` function on the query:
```
query.from("questions")
query.where("text", FirestoreQuery.OPERATOR.EQUAL, "What's your name?")
```
Anyway indexing should be always needed for any query regarding collections. The error you are getting should also print in console a link. If you click it (or copy/paste it) it will navigate you to the page where you can set the indexes.

**Reply by pndalal** [reply]
Oops! My code does contain "where", but I think in copy-pasting it here I messed up. Good catch, I've edited the previous comment and post.

Yes, I see the link, and it does fix the problem. It just doesn't make sense to me why I need a collection GROUP index, when my query pertains to a single collection which should be covered by the Automatic Index Settings. In other words, I shouldn't have to manually create any new indexes for such a simple query - at least that's my understanding. But I'm a Firestore beginner, so yeah.

**Reply by fenix-hub** [reply]
Oh well, to this question, sadly it's not something we can give you an exact answer about. It's just how Firestore works and it is unrelated to our plugin since it is not something we can control.
To make it simple, indexing is needed by default because it is how Firebase is able to assure you high performances for any query. 
This happens because when you send a query, it is not really like an SQL query : Firebase is not going to actually parse line per line the query and look for your resources as expected by SQL, but it's going to retrieve some resources related to a "topic" (index) which is linked to your query. 
To get a deeper insight fyk:
https://firebase.google.com/docs/firestore/query-data/index-overview
https://firebase.google.com/docs/firestore/query-data/indexing

Hope this answer helps you understanding this.

**Reply by pndalal** [reply]
I do understand that indexes are required by Firestore, but I'm wondering why the COLLECTION_GROUP_ASC index _specifically_ is required. From what I'm reading, it should NOT be required for the basic query I'm doing.

In the second link you provided, it says: "The indexes required for the most basic queries are [automatically created](https://firebase.google.com/docs/firestore/query-data/index-overview#single-field-indexes) for you."

There is a section on [Automatic indexing](https://firebase.google.com/docs/firestore/query-data/index-overview?authuser=0#automatic_indexing) where it says: "For each non-array and non-map field, Cloud Firestore defines two collection-scope single-field indexes, one in ascending mode and one in descending mode."

This means that the default indexes provided by Firestore should be sufficient for my case and that I don't need to create a separate COLLECTION_GROUP_ASC index. It could be that I've misconfigured something on the database end, but I'm not sure that's the case, because I've reproduced this issue with the most basic query on the simplest data. Which is why it seems like there's something pertaining to the use of the plugin.

**Reply by pndalal** [reply]
Figured it out. I didn't realize there was a second optional parameter in the "from". See my [comment ](https://github.com/GodotNuts/GodotFirebase/discussions/234#discussioncomment-1372127) above to see the working code.

**Reply by fenix-hub** [reply]
Oh sorry - maybe there was some misunderstanding from me. I thought that your call was intentional, maybe I should add this to the documentation so it is more clear.
As you can see from [here](https://firebase.google.com/docs/firestore/reference/rest/v1/StructuredQuery#collectionselector) the `all_descendants` parameter is used to issue queries on deeper level collections - that's why you need to create the index, which is always required.

**Reply by pndalal** [reply]
No worries. If I'd had more experience with Firestore I would have been able to make my question more clear, but for me, all of this is new. I appreciate your quick responses and support!

**Reply by fenix-hub** [reply]
It's ok, no one is born expert. Feel free to open new discussions or reach us out on Discord if you need help both on Firebase or just Godot :)
Glad we helped you.
